lib/Sema/TypeChecker.cpp

```diff
@@ -573,6 +573,77 @@ void swift::performWholeModuleTypeChecking(SourceFile &SF) {
 #endif
 }
 
+void swift::checkInconsistentImplementationOnlyImports(ModuleDecl *MainModule) {
+  bool hasAnyImplementationOnlyImports =
+      llvm::any_of(MainModule->getFiles(), [](const FileUnit *F) -> bool {
+    auto *SF = dyn_cast<SourceFile>(F);
+    return SF && SF->hasImplementationOnlyImports();
+  });
+  if (!hasAnyImplementationOnlyImports)
+    return;
+
+  auto diagnose = [MainModule](const ImportDecl *normalImport,
+                               const ImportDecl *implementationOnlyImport) {
+    auto &diags = MainModule->getDiags();
+    {
+      InFlightDiagnostic warning =
+          diags.diagnose(normalImport, diag::warn_implementation_only_conflict,
+                         normalImport->getModule()->getName());
+      if (!normalImport->getAttrs().isEmpty()) {
+        warning.fixItInsert(normalImport->getStartLoc(),
```

Comment:
I don't think you test this fix-it, and I'm not sure what the `if` statement is for. What's the story here? Should it be told in a comment?

Reply:
It's a good thing you said this because I got the condition backwards. ><